overlord/state/state.go

```diff
@@ -482,6 +491,58 @@ func (s *State) GetMaybeTimings(timings interface{}) error {
 	return nil
 }
 
+// AddTaskStatusChangedObserver adds a callback function that will be invoked
+// whenever tasks change status.
+// NOTE: Callbacks registered this way may be invoked in the context
+// of the taskrunner, so the callbacks should be as simple as possible, and return
+// as quickly as possible, and should avoid the use of i/o code or blocking, as this
+// will stop the entire task system.
+func (s *State) AddTaskStatusChangedObserver(f func(t *Task, old, new Status)) int {
+	s.reading()
+	id := s.lastObserverId
+	s.lastObserverId++
+	s.taskObservers[id] = f
+	return id
+}
+
+func (s *State) RemoveTaskStatusChangedObserver(observerId int) {
+	s.reading()
+	delete(s.taskObservers, observerId)
+}
+
+func (s *State) notifyTaskStatusChangedObservers(t *Task, old, new Status) {
+	s.reading()
+	for _, f := range s.taskObservers {
+		f(t, old, new)
+	}
+}
+
+// AddChangeStatusChangedObserver adds a callback function that will be invoked
+// whenever a Change changes status.
+// NOTE: Callbacks registered this way may be invoked in the context
+// of the taskrunner, so the callbacks should be as simple as possible, and return
+// as quickly as possible, and should avoid the use of i/o code or blocking, as this
+// will stop the entire task system.
+func (s *State) AddChangeStatusChangedObserver(f func(chg *Change, old, new Status)) int {
```

Comment:
same about the return argument: (id int)

Reply:
done!